### PR TITLE
JS Components: Add IconTooltip component

### LINKED ISSUE #25914: JS Components: Add a Tooltip component
Reference: `p1HpG7-hvT-p2`

Add a Tooltip component, like in the screenshot, which will be used in quite a few places across projects. Please ensure we could preview it with storybook.

Image

### PR DESCRIPTION
**Pull request by dognose24**
This PR adds an IconTooltip component to Jetpack reusable components.

Fixes #25914 

#### Changes proposed in this Pull Request:

* Add `IconTooltip` component reusing [Gutenberg `Popover` component](https://github.com/WordPress/gutenberg/tree/trunk/packages/components/src/popover)
* Add storybook for `IconTooltip` component
* Ensure the component takes props: `placement`, `animate`, `iconCode`, `title` and `children`
截圖 2022-09-07 下午4 15 27

#### Other information:

- [x] Have you written new tests for your changes, if applicable?
- [x] Have you checked the E2E test CI results, and verified that your changes do not break them?

#### Jetpack product discussion

N/A
#### Does this pull request change what data or activity we track or use?

N/A
#### Testing instructions:

* Run `cd projects/js-packages/storybook && npm run storybook:dev` and confirm that you can see `IconTooltip` component in Storybook.
* Under `JS Packages` → `Components` → `IconTooltip`, choose the story `Has Content` and try changing some of the controls. You should be able to adjust the placement, animate, icon code, and tooltip title.
* `Hover` on the icon and you should see the tooltip panel at the set placement.

Note: The tooltip could change the placement automatically when there is no space enough to show the tooltip at the set placement, which is made by the [origin Popover component](https://github.com/WordPress/gutenberg/tree/trunk/packages/components/src/popover#flip-boolean).

### CODE REVIEW

**Comment by a8ck3n** [comment]
I'm seeing an issue in testing where the visible state gets mixed up. This video shows what I mean:

https://user-images.githubusercontent.com/40267301/189049096-637cc8d0-b692-4d81-a966-04e308034b72.mp4

If you have it set to animate and hover from the bottom of the icon, there's a quick flash where the animation starts and then disappears. From then on, the state is reversed. I wasn't able to reproduce this with animation disabled.

I'm wondering if this has something to do with line 37. Do we want to look at the previous state here?

**Review by a8ck3n — DISMISSED** [review]
This works well but I'm seeing an issue with the show/hide state getting mixed up. See the video for an example of what I'm seeing.

**Comment by dognose24** [comment]
Thanks for the quick review, @a8ck3n! The state should be set to the exact value to avoid unpaired operations, which could happen in subtle moving.

File: projects/js-packages/components/components/icon-tooltip/index.tsx [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+	};
+
+	const args = {
+		iconCode,
+		children: (
+			<div>
+				<div className="icon-tooltip-title">{ title }</div>
+				<div className="icon-tooltip-content">{ children }</div>
+			</div>
+		),
```
Comment (a8ck3n) [inline projects/js-packages/components/components/icon-tooltip/index.tsx]: I find this a bit confusing having a children prop and then a secondary children grouping that combines title & children. Maybe we could rename the prop?

File: projects/js-packages/components/components/icon-tooltip/stories/index.tsx [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+			control: { type: null },
+		},
+	},
+};
+
+const Template = args => (
+	<div style={ { position: 'absolute', height: '2000px', left: '400px', top: '400px' } }>
+		<IconTooltip { ...args } />
+	</div>
+);
```
Comment (a8ck3n) [inline projects/js-packages/components/components/icon-tooltip/stories/index.tsx:46]: The styling on this DIV doesn't play nice with Storybook. For example, the controls are offscreen in the Docs view or on a smaller screen. Could we maybe position this closer to the top-left of the canvas?

**Review by a8ck3n — COMMENTED** [review]
This is working pretty good for me. I'm no longer seeing the visibility bug that I reported previously. I left a few comments inline that aren't really blockers.

The one thing I'd ask is that we document the props a bit. I'm not exactly sure what many of them do specifically without digging into the code.

For a final approval, as I'm not super confident around all the bits in here, maybe we can ping @kangzj for a once over.

File: projects/js-packages/components/components/icon-tooltip/index.tsx [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+			<span
+				style={ { cursor: 'pointer' } }
+				onMouseEnter={ createToggleIsOver( 'onMouseEnter', true ) }
+				onMouseLeave={ createToggleIsOver( 'onMouseLeave' ) }
+			>
+				<Gridicon className={ iconClassName } icon={ args.iconCode } size={ 18 } />
+			</span>
+
+			<div className="icon-tooltip-helper">
+				{ isVisible && <Popover { ...args }>{ args.children }</Popover> }
```
Comment (oskosk) [inline projects/js-packages/components/components/icon-tooltip/index.tsx]:
```suggestion
				{ isVisible && (
					<Popover { ...args }>
						<div>
							<div className="icon-tooltip-title">{ title }</div>
							<div className="icon-tooltip-content">{ children }</div>
						</div>
					 </Popover>
				) }
```

File: projects/js-packages/components/components/icon-tooltip/index.tsx (hunk 1 above)
Comment (oskosk) [inline projects/js-packages/components/components/icon-tooltip/index.tsx]:
```suggestion
```

I think some complexity can be reduced by moving the markup here into the returned markup section below.

**Review by oskosk — COMMENTED** [review]
Left a suggestion after reading @a8ck3n's comment on the `children` prop

File: projects/js-packages/components/components/icon-tooltip/stories/index.tsx (hunk 2 above)
Comment (dognose24) [reply projects/js-packages/components/components/icon-tooltip/stories/index.tsx:46]: Yes, @a8ck3n! The original intention of leaving these styles is to ensure the `position` in the Storybook will work as the player set. I would move it to the more appropriate edge.

File: projects/js-packages/components/components/icon-tooltip/index.tsx (hunk 1 above)
Comment (dognose24) [reply projects/js-packages/components/components/icon-tooltip/index.tsx]: Yes, @a8ck3n! The prop `children` is made for the tooltip body content. I would refactor the usage of `title` and `children` to a more explicit structure for intuition, as [what @oskosk suggested](https://github.com/Automattic/jetpack/pull/26081#pullrequestreview-1103140216).

File: projects/js-packages/components/components/icon-tooltip/index.tsx [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+			} else {
+				setIsVisible( _isVisible );
+			}
+		};
+	};
+
+	const args = {
+		iconCode,
+		// To be compatible with deprecating option `position`
+		position: placementsToPositions( placement ),
```
Comment (kangzj) [inline projects/js-packages/components/components/icon-tooltip/index.tsx:62]: Can you tell me more about this? I suppose if an attribute is deprecated, we could make it transparent?

File: projects/js-packages/components/components/icon-tooltip/index.tsx [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+		placement,
+		animate,
+		noArrow: false,
+		resize: false,
+		flip: false,
+		offset: 10,
```
Comment (kangzj) [inline projects/js-packages/components/components/icon-tooltip/index.tsx]: It'll be great if magic numbers are documented 😄

**Review by kangzj — DISMISSED** [review]
I'm happy for it to merge as is. Left a couple of comments no blockers 👍

File: projects/js-packages/components/components/icon-tooltip/index.tsx (hunk 1 above)
Comment (oskosk) [reply projects/js-packages/components/components/icon-tooltip/index.tsx]: Addressed in d6525d20682a6951a47270514d06df545fcf8226

File: projects/js-packages/components/components/icon-tooltip/index.tsx (hunk 4 above)
Comment (dognose24) [reply projects/js-packages/components/components/icon-tooltip/index.tsx:62]: Since [`placement` is announced](https://github.com/WordPress/gutenberg/tree/trunk/packages/components/src/popover#placement-top--top-start--top-end--right--right-start--right-end--bottom--bottom-start--bottom-end--left--left-start--left-end) as the replacement of legacy `position` in `Popover` component, I tend to use the newer prop and implicitly make it compatible with older version dependencies in this `IconTooltip` component. Even though the [newer version `Popover` component has processed the legacy prop](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/popover/index.tsx#L244-L246) `position`.

In other words, users can only take care of `placement` no matter [what version of components](https://www.npmjs.com/package/@wordpress/components) they are using.

**Review by kangzj — DISMISSED** [review]
LGTM

I have suggestion but which could totally be done in a followup PR.

image

When the tooltip is up, user might tend to move the mouse to the popup and click a link or select text, so could we change the behavior a bit possibly? - when user move mouse away from icon but still on the popup, we still keep the popup open. when moved away from popup and the icon, then close the popup. Or even better, allow devs to choose the behavior.

**Comment by oskosk** [comment]
Pushed rebase in 655eed45a43ac1124c71a9f181c160d4354cbb5e and versions update in e001eb9375758e252f8259a517c602a4ca9fcb31

**Review by oskosk — APPROVED** [review]
LGTM!

Renato left a comment for adapting this component in the future to adapt it to some practices that have been adopted for other components in https://github.com/Automattic/jetpack/issues/26207

**Comment by kraftbj** [comment]
r252613-wpcom
